tests/common/dualtor/dual_tor_io.py

```diff
@@ -0,0 +1,424 @@
+        """
+        This method examines packets collected by sniffer thread
+        The method compares TCP payloads of the packets one by one (assuming all payloads are consecutive integers),
+        and the losses if found - are treated as disruptions in Dataplane forwarding.
+        All disruptions are saved to self.lost_packets dictionary, in format:
+        disrupt_start_id = (missing_packets_count, disrupt_time, disrupt_start_timestamp, disrupt_stop_timestamp)
+        """
+        def examine_each_packet(dut_mac, packets):
+            lost_packets = dict()
+            prev_payload, prev_time = 0, 0
+            sent_payload = 0
+            disruption_start, disruption_stop = None, None
+            received_counter = 0    # Counts packets from dut.
+            for packet in packets:
+                if packet[scapyall.Ether].dst == dut_mac:
+                    # This is a sent packet - keep track of it as payload_id:timestamp.
+                    sent_payload = int(str(packet[scapyall.TCP].payload).replace('X',''))
+                    sent_packets[sent_payload] = packet.time
+                    continue
+                if packet[scapyall.Ether].src == dut_mac:
+                    # This is a received packet.
+                    received_time = packet.time
+                    received_payload = int(str(packet[scapyall.TCP].payload).replace('X',''))
+                    received_counter += 1
+                if not (received_payload and received_time):
+                    # This is the first valid received packet.
+                    prev_payload = received_payload
+                    prev_time = received_time
+                    continue
+                if received_payload - prev_payload > 1:
+                    # Packets in a row are missing, a disruption.
+                    lost_id = (received_payload -1) - prev_payload # How many packets lost in a row.
+                    disrupt = (sent_packets[received_payload] - sent_packets[prev_payload + 1]) # How long disrupt lasted.
+                    # Add disrupt to the dict:
+                    lost_packets[prev_payload] = (lost_id, disrupt, received_time - disrupt, received_time)
+                    logger.info("Disruption between packet ID %d and %d. For %.4f " % (prev_payload, received_payload, disrupt))
+                    if not disruption_start:
+                        disruption_start = datetime.datetime.fromtimestamp(prev_time)
+                    disruption_stop = datetime.datetime.fromtimestamp(received_time)
+                prev_payload = received_payload
+                prev_time = received_time
+            if received_counter == 0:
+                logger.error("Sniffer failed to filter any traffic from DUT")
+            else:
+                logger.info("Total number of filtered incoming packets captured {}".format(received_counter))
+            if lost_packets:
+                logger.info("Disruptions happen between %s and %s." % \
```

Comment:
Looks like we are just logging the last disruption to happen, any way to capture all of them?

Reply:
All disruptions (between consecutive packets) are already being captured inside the packet loop above. This log is the high-level loss to understand if the start and end range corresponds to the timestamp between which action is performed.